specification/under_development/profiles/Adaptive_Profile.md

```diff
@@ -0,0 +1,988 @@
+chosen by the backend and is identified by a string label in the produced
+schema. Each output schema contains sufficient information to allow quantum
+programming frameworks to generate a user-friendly presentation of the returned
+values in the requested order, such as, e.g., a histogram of all results when
+running the program multiple times.
+
+**Bullet 4: Conditional Forward branching** <br />
+The Adaptive Profile can allow for arbitrary forward branching with the restriction being that
+branch instructions cannot express programs that do not terminate. By default, support for
+backwards branching is not a requirement, though a back-end can opt into **Bullet 13** to support
+backwards branching. In the case that **Bullet 13** is not opted into, then a simple static analysis
+can check for cycles in the control flow graph and reject any program with a cycle as invalid.
+Depending on the vector of capabilities defined by an Adaptive Profile program 
+(like when **Bullet 13** is opted into), 
+proving non-termination with a static analysis may impossible due to the capabilities
+forming a Turing-complete subset of LLVM. Therefore, it is a rule that a *valid* Adaptive Profile
+program should not contain non-terminating loops and that a front-end generating a program with a 
+non-terminating loop is generating a program that is not compliant with the Adaptive Profile specification.
+However, since there is no static analysis that can prove termination then it is up to back-ends to enforce 
+termination guarantees via a means of their choosing (for example a watchdog process
+with a timeout that will kill an executing Adaptive Profile program if it takes
+too much time). The following forms of branch instructions can be supported in the 
+profile, but at least the basic `br` instruction must be supported from the list below :
+| LLVM Instruction         | Context and Purpose                                                                              | Rules for Usage                                                                                             |
+| :----------------------- | :----------------------------------------------------------------------------------------------- | :---------------------------------------------------------------------------------------------------------- |
+| `br i1 label %true_branch, label %false_branch`                   | Branch to one of two target blocks. | Cannot be used to implement a non-terminating loop.     |
+| `switch i64 %reg, label %otherwise [i64 0, %zero_branch i64 1, %one_branch ... i64 n, %n_branch]`                     | Used to branch to one of many objects based on an integer key.                            | Cannot be used to implement a non-terminating loop. |
+| `select i1 %0, i64 %true_reg, i64 %false_reg`                    | conditionally select the value in a register. | Cannot be used to implement a non-terminating loop.                            |
+
```

Comment:
It seems like support for the `switch` and `select` instructions is optional, but since they are not part of the optional Bullets of the Adaptive Profile (i.e. Bullets 8-15), there is no way to specify them in the module flags like the other optional features/instructions of the adaptive profile. It might be more clear to specify that support for all 3 is required, or else place support for the optional `switch` and `select` instructions into a new module flag.

Reply:
Hmm this is a good point. Currently, Quantinuum supports `select` and `br` but not `switch`, however I think `switch` support could be very useful for reasoning about control flow independence and doing better static analysis for back-end optimization.  I don't see why a back-end supporting `br` couldn't support the other two so it may still make sense to have all of these be mandatory instructions to implement. 

But like splitting classical compute support into different buckets with their own module flags, I think it's probably a good idea for `br` to be the mainline support instruction and for `select` and `switch` to be optional instructions to support.

Reply:
From a validation perspective, it would be easier if support for `select` and `switch` were optional under their own distinct option. That way, validation tools can check against the corresponding module attribute.

Reply:
Of note, while the `switch` instruction is certainly optional in LLVM (and there even exists a transformation pass, [lowerswitch](https://www.llvm.org/docs/Passes.html#lowerswitch-lower-switchinsts-to-branches), that will turn the single switch instruction into the equivalent branching), the `select` instruction is much harder to eliminate. Either `select` needs to be supported or empty branches have to be allowed and lots of care taken with optimization passes that will aggressively turn those empty branches back into `select`. For example, consider this pseudo-code:
```
c = 0;
if (condition) {
    c = 1;
} else {
    c = 2;
}
```
Using just branching and phi-nodes, that would turn into something like:
```llvm
br i1 %condition, label %then-branch, label %else-branch

then-branch:
br label %fallthrough

else-branch:
br label %fallthrough

fallthrough:
%c = phi i64 [ 1, %then-branch ], [ 2, %else-branch ]
```
Both blocks used by the branch instruction are completely empty because the phi-node does all the work of choosing a value based on control flow. This is exactly what `select` is for, so any cfg simplification pass ends up turning this into:
```llvm
%c = select i1 %condition, i64 1, i64 2
```
This requires a custom pass to undo it, which would need to be run after any other optimizations to make sure they don't reintroduce the `select`.

Reply:
Based on this, one course of action is the following:
- Make `br` the only required instruction for Bullet 4 (Conditional Forward Branching), and Bullet 15 (Backwards Branching).
- Since both the `phi` and `select` instructions are used in the context of manipulating integer, floating point and fixed-point calculations, the inclusion of these two instructions can be tied to the computation options (Bullets 9, 10 and 11).
- For simplicity, exclude the `switch` instruction from this version of the spec since it is not required for any of the bullet points to work, and it easy to replace. There is an out-of-the-box [transformation pass](https://www.llvm.org/docs/Passes.html#lowerswitch-lower-switchinsts-to-branches) that rewrites switch instructions with a sequence of branches.

Reply:
I really like that course of action Cesar--maybe with the caveat that the `switch` instruction still becoming an optional feature. From my perspective long term, I eventually want adaptive profile programs to support switch instructions because it helps make it more easily recognizable. I'll update based on these points.

Reply:
@cesarzc and I spoke a bit about this, and I think we missed something on this one. The need for `phi` and `select` is not specific to integers, floating point and fixed point. Because the required elements allow for dynamic Boolean values, and those values can be combined or branched upon by complex logic, it may still be necessary to use a `phi` or `select` to decide between two different values of a dynamic Boolean for later branching. I think that means `phi` and `select` need to be part of the required instruction support list.